Lib/asyncore.py

```diff
@@ -399,6 +403,9 @@ def close(self):
                 if why.args[0] not in (ENOTCONN, EBADF):
                     raise
 
+    def fileno(self):
+        return self.socket.fileno()
```

Comment:
dispatcher and file_wrapper try to behave like a socket, so using using socket.fileno() seems like the best approach, and it is also compatible with older code that was calling dispatcher.fileno() and dispatches.socket.fileno().

I think we can leave file_wrapper as is for now, since the patch fixes the issue without this change.

So the only issue in this version of the patch is calling fileno() when dispatcher.socket is None. Not sure that this is a real issue, it means that the instance is not created or configured properly.

Regarding setting self.fd to -1 - this is a bug in 2.7 that should be fixed.